tendermint/src/serializers.rs

```diff
@@ -1,71 +1,199 @@
+//! Vec<u8>              <-> HexString:            #[serde(with="serializers::bytes::hexstring")]
+//! Vec<u8>              <-> Base64String:         #[serde(with="serializers::bytes::base64string")]
+//! Vec<u8>              <-> String:               #[serde(with="serializers::bytes::string")]
+//!
+//! Any type that has the "FromStr" trait can be serialized into a string with serializers::primitives::string.
+//!
 
 use crate::account::{Id, LENGTH};
 use crate::{block, Hash, Signature};
 use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 use std::str::FromStr;
-use std::time::Duration;
-use subtle_encoding::{base64, hex};
 
-/// Parse `i64` from a JSON string
-pub(crate) fn parse_i64<'de, D>(deserializer: D) -> Result<i64, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    String::deserialize(deserializer)?
-        .parse::<i64>()
-        .map_err(|e| D::Error::custom(format!("{}", e)))
-}
+/// Serialize/deserialize primitive types (i64, u64, etc)
+pub mod primitives {
 
-/// Serialize `i64` as a JSON string
-#[allow(clippy::trivially_copy_pass_by_ref)]
-pub(crate) fn serialize_i64<S>(value: &i64, serializer: S) -> Result<S::Ok, S::Error>
-where
-    S: Serializer,
-{
-    format!("{}", value).serialize(serializer)
-}
+    /// Serialize into string, deserialize from string
+    pub mod string {
+        use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 
-/// Parse `u64` from a JSON string
-pub(crate) fn parse_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    String::deserialize(deserializer)?
-        .parse::<u64>()
-        .map_err(|e| D::Error::custom(format!("{}", e)))
-}
+        /// Deserialize string into T
+        pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
+        where
+            D: Deserializer<'de>,
+            T: std::str::FromStr,
+            <T as std::str::FromStr>::Err: std::fmt::Display,
+        {
+            String::deserialize(deserializer)?
+                .parse::<T>()
+                .map_err(|e| D::Error::custom(format!("{}", e)))
+        }
 
-/// Serialize `u64` as a JSON string
-#[allow(clippy::trivially_copy_pass_by_ref)]
-pub(crate) fn serialize_u64<S>(value: &u64, serializer: S) -> Result<S::Ok, S::Error>
-where
-    S: Serializer,
-{
-    format!("{}", value).serialize(serializer)
+        /// Serialize from T into string
+        pub fn serialize<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+            T: std::fmt::Display,
+        {
+            format!("{}", value).serialize(serializer)
+        }
+    }
 }
 
-/// Parse `Duration` from a JSON string containing a nanosecond count
-pub(crate) fn parse_duration<'de, D>(deserializer: D) -> Result<Duration, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    // TODO(tarcieri): handle 64-bit overflow?
```

Comment:
Is this handled by the new code or is nothing to be concerned about?

Reply:
It's handled the same way, I think Tony left a comment here because he wasn't sure if it's going to cause a problem. (There's not much info about what he meant here.)

In both implementations, the deserialization fails if the incoming JSON object contains a bigger number that fits `u64`. Since it's in nanoseconds and the max value for u64 is `18446744073709551615`, the max number representable is over 584 years. Since BFT time can be faster than real time, this is not really until 2554, it can happen earlier. If/when it happens, my guess is that the chain halts.

Simple workaround is to use u128 in our code but I have a feeling this will be a problem in the Go code too.

Sounds like low priority for me. How about we open an issue about it?

Reply:
> Sounds like low priority for me. How about we open an issue about it?

OK

Reply:
Opened issue #262 about this.